notion-distcopy/src/main/scala/com/ambiata/notion/distcopy/FileDistCopy.scala

```diff
@@ -0,0 +1,92 @@
+   *
+   * Don't copy files which have already been copied
+   */
+  def distCopyDownloadDirectory(from: S3Prefix, to: Location, locationIO: LocationIO): RIO[Unit] =
+    for {
+      toUri         <- hadoopPath(to)
+      fromAddresses <- from.listAddress.execute(locationIO.s3Client)
+      mappings      <- fromAddresses.traverseU(createDownloadMapping(HdfsLocation(toUri), locationIO))
+      _             <- DistCopyJob.run(Mappings(mappings.toVector.flatten), distCopyConfiguration(locationIO))
+    } yield ()
+
+  /** upload a large file by doing a distcopy from hdfs to s3 */
+  def distCopyUploadFile(from: Location, to: S3Pattern, locationIO: LocationIO): RIO[Unit] =
+    for {
+      fromUri <- hadoopPath(from)
+      _       <- DistCopyJob.run(Mappings(Vector(UploadMapping(new Path(fromUri), S3Address(to.bucket, to.unknown)))), distCopyConfiguration(locationIO))
+    } yield ()
+
+  /** upload a large file by doing a distcopy from hdfs to s3 */
+  def distCopyUploadDirectory(from: Location, to: S3Location, locationIO: LocationIO): RIO[Unit] =
+    for {
+      fromUri   <- hadoopPath(from)
+      fromPaths <- Hdfs.globFiles(new Path(fromUri, "*")).run(locationIO.configuration)
+      mappings  <- fromPaths.traverseU(createUploadMapping(to, locationIO))
+      _         <- DistCopyJob.run(Mappings(mappings.toVector.flatten), distCopyConfiguration(locationIO))
+    } yield ()
+
+  /** @return true if the path exists */
+  def pathExist(path: Path, locationIO: LocationIO): RIO[Boolean] =
+    Hdfs.exists(path).run(locationIO.configuration)
+
+  /** @return true if the address exists */
+  def addressExist(address: S3Address, locationIO: LocationIO): RIO[Boolean] =
+    address.exists.execute(locationIO.s3Client)
+
+  /** @return a configuration for dist copy based on the current LocationIO configuration */
+  def distCopyConfiguration(locationIO: LocationIO): DistCopyConfiguration =
+    DistCopyConfiguration.Default.copy(hdfs = locationIO.configuration, client = locationIO.s3Client)
+
+  /** create a Download mapping for a file if it doesn't exist */
+  def createDownloadMapping(to: HdfsLocation, locationIO: LocationIO)(address: S3Address): RIO[Option[DownloadMapping]] = {
+    val toPath = new Path((to.dirPath <|> FilePath.unsafe(address.key).basename).path)
+    pathExist(toPath, locationIO).map(exists => if (exists) None else Some(DownloadMapping(address, toPath)))
+  }
+
+  /** create an Upload mapping for a file if it doesn't exist */
+  def createUploadMapping(to: S3Location, locationIO: LocationIO)(path: Path): RIO[Option[UploadMapping]] = {
+    val toAddress = S3Prefix(to.bucket, to.key) | path.getName
+    addressExist(toAddress, locationIO).map(exists => if (exists) None else Some(UploadMapping(path, toAddress)))
+  }
+
+  /**
+   * extract a URI for local and hdfs locations
+   * in order to pass it to a map-reduce job
+   */
+  def hadoopPath(location: Location): RIO[String] =
```

Comment:
Do you mean that `def hadoopPath(location: Location): RIO[HdfsLocation]` or `def hadoopPath(location: Location): String \/ HdfsLocation` would be more appropriate?